### ISSUE TITLE
Add ability to customize structure of error messages

### ISSUE THREAD

**Issue by Cellane** [vapor/vapor#1769]
I might have missed some obvious way of doing this but it seems to me that currently, there’s no way to customize the structure of error messages sent by Vapor. When throwing an `Abort`, you can, of course, provide a `reason` that will be put into the `error` field, but almost anything beyond that seems impossible. That makes it difficult to accomplish, say, a following error structure:

```json
{
    "error": "The object already exists",
    "code": 4242,
    "seeAlso": [
        42
    ]
}
```

I’m sorry if I missed a way to do this in Vapor already! I attempted to find a solution on Vapor Discord but was unable to reach an easy solution even with the help of others.

### Environment

* Vapor Framework version: 3.1.7
* Vapor Toolbox version: 3.0.6
* OS version: 10.13.5

**Comment by tanner0101** [comment]
Yeah that is a limitation of `ErrorMiddleware`. The good news is `ErrorMiddleware` is totally optional and you can replace it with a more powerful implementation if you want. 

I do think we could / should make the default one more powerful though. Right now the response that `ErrorMiddleware` produces is the nested struct `ErrorResponse`. An interesting solution to this might be checking if the `Error` caught conforms to something like `ResponseEncodable` (or `Content` or `Codable`) and in that case, just serializing the `Error` as a response directly instead of trying to transform it to `ErrorResponse`.

**Comment by MrLotU** [comment]
@Cellane With your permission I’d like to close this in favor of #1875 and get an implementation of that in Vapor 4 for sure, and maybe a backport to V3
Cc: @tanner0101

**Comment by Cellane** [comment]
I don’t have a problem with that but I kind of think this ticket should probably be closed only after the new implementation is fully merged.

**Comment by MrLotU** [comment]
Sounds good. Would you be able to edit your original issue a little so it more reflects the approach in #1875 using `ResponseEncodable`?

**Comment by MrLotU** [comment]
Closed per merge of #1875
